mlir/lib/Conversion/FuncToEmitC/FuncToEmitCPass.cpp

```diff
@@ -0,0 +1,49 @@
+//===- FuncToEmitC.cpp - Func to EmitC Pass ---------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a pass to convert the Func dialect to the EmitC dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/FuncToEmitC/FuncToEmitCPass.h"
+
+#include "mlir/Conversion/FuncToEmitC/FuncToEmitC.h"
+#include "mlir/Dialect/EmitC/IR/EmitC.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir {
+#define GEN_PASS_DEF_CONVERTFUNCTOEMITC
+#include "mlir/Conversion/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+
+namespace {
+struct ConvertFuncToEmitC
+    : public impl::ConvertFuncToEmitCBase<ConvertFuncToEmitC> {
+  void runOnOperation() override;
+};
+} // namespace
+
+void ConvertFuncToEmitC::runOnOperation() {
+  ConversionTarget target(getContext());
+
+  target.addLegalDialect<emitc::EmitCDialect>();
+  target.addIllegalOp<func::CallOp>();
+  target.addIllegalOp<func::FuncOp>();
+  target.addIllegalOp<func::ReturnOp>();
+
+  RewritePatternSet patterns(&getContext());
+  populateFuncToEmitCPatterns(patterns);
```

Comment:
Could we move this to the initialize class member?

Reply:
Technically yes, but it seems uncommon. Most conversation passes call a function to populate the patterns within `runOnOperation()`. The only pass I am aware of which does this in the constructor is the TOSA validation pass: https://github.com/llvm/llvm-project/blob/121a0ef0bbc1a69760b69391b387e2efda0ad922/mlir/lib/Dialect/Tosa/Transforms/TosaValidation.cpp#L98-L100

Since it seems not common in the code base, I would leave it as is.